### PR TITLE
Feature/client certificates

### LINKED ISSUE #2569: Add config to specify client certificate to connect to SSL server
Conan uses the "cacert" file installed in the conan home so you can add new authorities to verify the certificate from a server.

But some users try to use a client certificate, so the server validates the connecting client. (reverse mechanism)
We need to add a new config entry to point to a file containing the [user certificate](http://blog.nategood.com/client-side-certificate-authentication-in-ngi) and key.

In python requests it can be done passing the ["cert" parameter](http://docs.python-requests.org/en/master/user/advanced/#client-side-certificates), that can be a tuple of 2 elements, first certificate, and a key, or just a string with a (commonly pem) file containing both certificate and private key.

The feature is already tested against a local nginx.

### PR DESCRIPTION
**Pull request by lasote**
- [x] Refer to the issue that supports this Pull Request. Closes #2569 Probably improves (to confirm if it helps) #2315 by introducing a new config "no_proxy_match" matching fnmatch to exclude urls.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

https://github.com/conan-io/docs/pull/585/files

### CODE REVIEW

File: conans/client/rest/conan_requester.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import os
+
+from conans.util.files import save
+
+
+class ConanRequester(object):
+
+    def __init__(self, requester, client_cache):
+        self.proxies = client_cache.conan_config.proxies or {}
+        self.no_proxy_match = [el.strip() for el in
```
Comment (memsharded) [inline conans/client/rest/conan_requester.py]: private attribute ``_no_proxy_match``?

File: conans/client/rest/conan_requester.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            if os.path.exists(client_cache.client_cert_key_path):
+                # Requests can accept a tuple with cert and key, or just an string with a
+                # file having both
+                self._client_certificates = (client_cache.client_cert_path,
+                                             client_cache.client_cert_key_path)
+            else:
+                self._client_certificates = client_cache.client_cert_path
+
+    def _should_skip_proxy(self, url):
+        if not self.no_proxy_match:
```
Comment (memsharded) [inline conans/client/rest/conan_requester.py]: not strictly necessary, the for will not iterate already.

File: conans/client/rest/conan_requester.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        for entry in self.no_proxy_match:
+            if fnmatch.fnmatch(url, entry):
+                return True
+
+        return False
+
+    def _add_kwargs(self, url, kwargs):
+        if kwargs.get("verify", None) is True:
+            if not os.path.exists(self._client_cache.cacert_path):
```
Comment (memsharded) [inline conans/client/rest/conan_requester.py]: A bit ugly in this method, called all the time? Maybe the path once in the constructor? Maybe created by ``client_cache``?

File: conans/client/client_cache.py [hunk 4]
```diff
@@ -36,8 +44,15 @@ def __init__(self, base_folder, store_folder, output):
...
         self._default_profile = None
         self._no_lock = None
+        self.client_cert_path = normpath(join(self.conan_folder, CLIENT_CERT))
+        self.client_cert_key_path = normpath(join(self.conan_folder, CLIENT_KEY))
+
         super(ClientCache, self).__init__(self._store_folder)
 
+    @property
+    def cacert_path(self):
+        return os.path.join(normpath(join(self.conan_folder, CACERT_FILE)))
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/client_cache.py]: definitely a copy paste issue LoL

File: conans/client/rest/version_checker.py [hunk 5]
```diff
@@ -14,32 +14,29 @@ def __init__(self, requester, client_version, min_server_compatible_version, out
...
         self.min_server_compatible_version = min_server_compatible_version
 
-    def get(self, url, auth=None, headers=None, verify=None, stream=None):
-        headers = headers or {}
-        headers['X-Conan-Client-Version'] = str(self.client_version)
-        ret = self.requester.get(url, auth=auth, headers=headers, verify=verify, stream=stream)
+    def _add_version_header(self, kwargs):
+        if "headers" not in kwargs or kwargs["headers"] is None:
+            kwargs["headers"] = {}
+        kwargs["headers"]['X-Conan-Client-Version'] = str(self.client_version)
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/rest/version_checker.py]: I like the one line but also the method.

File: conans/test/functional/client_certs_test.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                                                        client.client_cache.client_cert_key_path))
+
+        # assert that the cacert file has not been created yet, but will when contacting a server
+        self.assertFalse(os.path.exists(client.client_cache.cacert_path))
+
+        test_server = TestServer([],  # write permissions
+                                 users={"lasote": "mypass"})  # exported users and passwords
+        servers = {"default": test_server}
+        client = TestClient(servers=servers, users={"default": [("lasote", "mypass")]})
+        client.run("install something_fake/1.0@conan/stable", ignore_error=True)
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/test/functional/client_certs_test.py]: Ok, but the goal is not to check that it fails, of course it fails, I'm installing a missing reference, I'm testing than even failing I get the cacert generated

File: conans/client/rest/version_checker.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/rest/version_checker.py]: One line is not valid, kwargs["headers"] can exist, but with None value

**Pull request merged**